internal/graph/check_test.go

```diff
@@ -2998,3 +3005,531 @@ func TestIteratorReadStartingFromUser(t *testing.T) {
 		})
 	}
 }
+
+func TestConsumeUsersets(t *testing.T) {
+	type dsResults struct {
+		tuples []*openfgav1.TupleKey
+		err    error
+	}
+	tests := []struct {
+		name                         string
+		model                        *openfgav1.AuthorizationModel
+		tuples                       []dsResults
+		usersetsChannelResult        []usersetsChannelStruct
+		ctxCancelled                 bool
+		expectedResolveCheckResponse *ResolveCheckResponse
+		errorExpected                error
+	}{
+		{
+			name: "userset_tuple_found",
+			model: parser.MustTransformDSLToProto(`
+				model
+					schema 1.1
+				type user
+				type group
+					relations
+						define member: [user]
+				type document
+					relations
+						define viewer: [group#member]`),
+			tuples: []dsResults{
+				{
+					tuples: []*openfgav1.TupleKey{
+						tuple.NewTupleKey("group:1", "member", "user:maria"),
+						tuple.NewTupleKey("group:2", "member", "user:maria"),
+						tuple.NewTupleKey("group:3", "member", "user:maria"),
+						tuple.NewTupleKey("group:4", "member", "user:maria"),
+					},
+					err: nil,
+				},
+			},
+			usersetsChannelResult: []usersetsChannelStruct{
+				{
+					err:            nil,
+					objectRelation: "group#member",
+					objectIDs:      []string{"0", "2", "8"},
+				},
+			},
+			ctxCancelled: false,
+			expectedResolveCheckResponse: &ResolveCheckResponse{
+				Allowed: true,
+				ResolutionMetadata: &ResolveCheckResponseMetadata{
+					DatastoreQueryCount: 1,
+				},
+			},
+			errorExpected: nil,
+		},
+		{
+			name: "userset_tuple_found_multi_batches",
+			model: parser.MustTransformDSLToProto(`
+				model
+					schema 1.1
+				type user
+				type group
+					relations
+						define member: [user]
+				type document
+					relations
+						define viewer: [group#member]`),
+			tuples: []dsResults{
+				{
+					tuples: []*openfgav1.TupleKey{
+						tuple.NewTupleKey("group:11", "member", "user:maria"),
+						tuple.NewTupleKey("group:12", "member", "user:maria"),
+						tuple.NewTupleKey("group:13", "member", "user:maria"),
+						tuple.NewTupleKey("group:14", "member", "user:maria"),
+					},
+					err: nil,
+				},
+				{
+					tuples: []*openfgav1.TupleKey{
+						tuple.NewTupleKey("group:11", "member", "user:maria"),
+						tuple.NewTupleKey("group:12", "member", "user:maria"),
+						tuple.NewTupleKey("group:13", "member", "user:maria"),
+						tuple.NewTupleKey("group:14", "member", "user:maria"),
+					},
+					err: nil,
+				},
```

Comment:
Is it on purpose that the three batches have the same tuples?

Reply:
Yes, this simulates the result returned by the iterator in response to each new batch from the usersetsChannel